File: btoa.c
/*
 * btoa: Convert text representing the binary of ascii characters.
 * Each binary sequence (in text) representing a character needs
 * to be delimited by a space ' ' or ended by a newline to be
 * properly printed.
 * Written by William S. Moore
 */
#include <stdio.h>
#include <ctype.h>

int main(int argc, char *argv[])
{
	if (argc > 1) {
		printf("btoa: Convert text representing the binary of ASCII characters.\n");
		printf("Note 1: Each binary sequence (in text) representing a character needs ");
		printf("to be delimited by a space ' ' or ended by a newline to be properly printed.\n");
		printf("You are reading this because arguments were passed to btoa.\n");
		return 1;
	}

	int c;
	int character_byte = 0;
	while((c = getchar()) != EOF) {
		// print stored byte if delimiter is found
		if (c == ' ' || c == '\n' || c == '\r') {
			if (isprint(character_byte) || isspace(character_byte)) {
				putchar(character_byte);
			}
			character_byte = 0;
			continue;
		}

		// check if c is 1 or 0 then shift into byte
		if (c != '0' && c != '1') {
			fprintf(stderr, "ERROR: Incompatible character!\n");
			return 1;
		}
		c = c - '0'; // convert to digit
		character_byte = (character_byte << 1) + c;
	}
	return 0;
}

File: BtoA.java
import java.io.IOException;

/**
 * btoa: Convert text representing the binary of ASCII characters.
 * Each binary sequence (in text) representing a character needs
 * to be delimited by a space ' ' or ended by a newline to be
 * properly printed.
 * * Ported from C by Gemini.
 */
public class BtoA {
    public static void main(String[] args) {
        // C logic: if (argc > 1)
        if (args.length > 0) {
            System.out.println("btoa: Convert text representing the binary of ASCII characters.");
            System.out.print("Note 1: Each binary sequence (in text) representing a character needs ");
            System.out.println("to be delimited by a space ' ' or ended by a newline to be properly printed.");
            System.out.println("Note 2: The newline character '\\n' is preserved in the output.");
            System.out.println("You are reading this because arguments were passed to btoa.");
            System.exit(1);
        }

        try {
            int c;
            int characterByte = 0;

            // System.in.read() returns -1 at the end of the stream, mimicking EOF
            while ((c = System.in.read()) != -1) {
                
                if (c == ' ') {
                    if (characterByte > 31) {
                        System.out.print((char) characterByte);
                    }
                    characterByte = 0;
                    continue;
                }

                if (c == '\n' || c == '\r') {
                    if (characterByte > 31) {
                        System.out.print((char) characterByte);
                    }
                    characterByte = 0;
                    System.out.print((char) c);
                    continue;
                }

                // Convert '0'/'1' characters to numeric 0/1 and shift into the byte
                int bit = c - '0';
                characterByte = (characterByte << 1) + bit;
            }
        } catch (IOException e) {
            System.err.println("Error reading input: " + e.getMessage());
        }
    }
}
